### PR TITLE
[issues/572] Extract inline CI shell scripts to standalone + bats tests

### LINKED ISSUE #572: Extract Inline CI Shell Scripts → Standalone + bats Tests
## Current State

20 inline shell scripts across 8 composite action files in `.github/actions/`. Most are trivial (one-liner `pnpm` calls), but 4-5 have non-trivial logic that we've already had bugs in.

The bugs we fixed in this issue all trace back to untestable inline scripts:
- `\d+(?= total)` matching Snapshots instead of Tests (no pipefail, wrong regex)
- `Tests:.*\K\d+` capturing trailing digit only (greedy `.*` + missing `\s`)
- `|| echo "0"` fallback silently broken (pipefail semantics)

Each bug was caught manually by reading CI output on a PR. bats tests would catch them locally in seconds.

## Reference: bats setup in couimet/my-claude-skills

- bats-core installed via Homebrew (`brew install bats-core`)
- Tests live in `tests/*.bats`, loaded via `load test_helper`
- Shared `test_helper.bash` creates/destroys `$TEST_TEMP_DIR` per test
- Scripts under test live in `skills/<name>/<script>.sh`
- Scripts use `#!/usr/bin/env bash`, `set -euo pipefail`, error code constants
- Run via `make test` (which runs `bats tests/`)
- Tests invoke scripts as child processes via `run "$SCRIPT"`, then assert `$status`, `$output`
- CI pipeline runs `make test` on every push — see `.github/workflows/test.yml` for the workflow template. Bringing this to RangeLink is a straightforward copy:

    ```yaml
    # .github/workflows/test-ci-scripts.yml
    name: CI Script Tests
    on:
      push:
        branches: [main]
      pull_request:
    permissions:
      contents: read
    jobs:
      test:
        runs-on: ubuntu-24.04
        timeout-minutes: 5
        steps:
          - uses: actions/checkout@v4
          - uses: bats-core/bats-action@4.0.0
            with:
              support-install: false
              assert-install: false
              detik-install: false
              file-install: false
          - run: bats tests/ci/
    ```

    The `bats-core/bats-action` GitHub Action installs bats on the runner — no Homebrew needed in CI. Same pattern as the reference project.

## PROS

- Catch regex/shell bugs locally in ~100ms instead of waiting 8 minutes for CI
- Tests document expected behavior — future contributors can see what the grep patterns should capture
- Follows the pattern already established in `my-claude-skills` (user familiarity)
- Scripts become reusable — could be invoked locally or in other workflows
- Eliminates the "edit YAML → push → wait for CI → read comment → fix" loop we just went through 3 times

## CONS

- New dev dependency: bats-core via Homebrew for local dev (`brew install bats-core`). CI uses `bats-core/bats-action` GitHub Action — no brew needed on runners
- Scripts currently reference `${{ inputs.* }}`, `${{ github.* }}`, `$GITHUB_OUTPUT`, `$GITHUB_STEP_SUMMARY` — these need to be abstracted as CLI arguments or env vars for testability
- CI integration overhead: composite actions must call the extracted script file, handle path resolution
- Not every script is worth extracting (one-liner `pnpm install` adds no value as a standalone file)
- Duplication: `run-integration-tests` and `run-integration-tests-with-extensions` share identical stats extraction — could share a script but that adds cross-action coupling

### Candidate A: `scripts/ci/extract-test-stats.sh` (highest priority)

Extracts from: `run-tests/action.yml` stats step + `run-integration-tests/action.yml` stats step + `run-integration-tests-with-extensions/action.yml` stats step

Currently duplicated across 3 files with minor variations. The script would:
- Read a test output file path as argument
- Accept `--mode jest|mocha` to pick the right grep patterns
- Output `PASSED=<N>` and `TOTAL=<N>` (for jest) or `PASSING=<N>` (for mocha) on stdout
- Default missing/unparseable values to 0
- Exit 0 on success, non-zero if file doesn't exist

This is the #1 priority — all 3 grep bugs we fixed live here. A bats test would exercise:
- Valid Jest output → correct passed/total
- Valid mocha output → correct passing count
- Empty/missing file → defaults to 0
- Output with Snapshots/Test Suites lines → only matches Tests: line
- Multiple packages in one file → picks last (largest) package

### Candidate B: `scripts/ci/extract-qa-tc-ids.py` (or .sh using awk/yq)

Extracts from: `post-pr-ci-summary/action.yml` lines 46-63

The inline Python with PyYAML dependency is the most fragile script in the pipeline. Extracting it lets us:
- Test with sample YAML fixtures (2 TCs, 6 TCs, empty, no automated=true)
- Test label filtering edge cases (empty label-filter, non-existent label, mixed labels)
- Catch the `is True` vs `== 'true'` semantics in tests
- Switch to awk/grep if we want to drop the PyYAML dependency

### Candidate C: `scripts/ci/build-summary-table.sh`

Extracts from: `post-pr-ci-summary/action.yml` lines 37-82

Takes title, duration, test counts, TC IDs as arguments; outputs a markdown table string. Testable in isolation for formatting edge cases (empty test counts, long TC ID list, special chars in title). However, the table building is straightforward string concatenation — lower ROI than A or B.

### Candidate D: `scripts/ci/count-todos.sh`

Extracts from: `check-todos/action.yml`

Has a potential bug (grepping directory listing instead of file contents for base branch comparison). Extraction would make this testable. But lower priority — not in the current issue scope.

## Suggested Directory Structure

```
.github/
  scripts/                    # extracted shell scripts
    ci/
      extract-test-stats.sh   # Candidate A
      extract-qa-tc-ids.py    # Candidate B
      build-summary-table.sh  # Candidate C
  actions/                    # composite actions (unchanged structure)
    run-tests/action.yml      # calls extract-test-stats.sh instead of inline
    ...

tests/                        # new: bats tests for CI scripts
  ci/
    extract-test-stats.bats
    extract-qa-tc-ids.bats
    ...
  test_helper.bash            # shared setup
```

## Recommendation

Start with Candidate A (`extract-test-stats.sh`) — it consolidates 3 nearly-identical inline scripts, covers all the grep bugs we fixed, and has the highest ROI. Candidate B second. Candidates C and D can be done opportunistically later.

This is orthogonal to the current issue — worth its own issue with a scoped plan rather than adding scope here.

### PR DESCRIPTION
**Pull request by couimet**
## Summary

Extracts the inline test-stats grep from composite actions into a standalone, tested shell script — the first batch of a broader cleanup spanning ~20 inline scripts across 8 action files. The inline Python in post-pr-ci-summary is absorbed into the existing `resolve-qa-labels.js` (which already parsed QA YAML), eliminating a duplicated YAML parser and removing the PyYAML dependency from CI.

Adds bats test infrastructure (`tests/shell/`, `pnpm test:bats`, CI workflow) so regex/shell bugs are caught locally in ~100ms instead of an 8-minute CI cycle. Remaining scripts and bats coverage for existing package-level scripts are deferred to follow-up issues (#574, #575, and more to come).

## Changes

- New `.github/scripts/ci/extract-test-stats.sh` — extracts Jest/mocha test stats via perl regex (cross-platform, no grep -P dependency). Mocha mode now extracts both passing and failed counts. Replaces duplicated inline grep in 3 composite actions.
- `resolve-qa-labels.js` extended with `--automated-only`, `--exclude-label`, `--format csv` flags and optional `--label`. Now serves both `test-release-run.sh` (label-based MOCHA_GREP) and `post-pr-ci-summary` (automated=true TC ID extraction). Removes the need for a separate Python script and its PyYAML dependency.
- `post-pr-ci-summary` action: added failed integration test count row and copyable re-run command block (extracted from the test report) to the PR comment. New `int-failed` and `report-file` inputs.
- New `tests/shell/` — bats test suite (12 tests) with shared `test_helper.bash` and fixture files for `extract-test-stats.sh`.
- New `.github/workflows/test-shell-scripts.yml` — runs `bats tests/shell/` on every push via `bats-core/bats-action@4.0.0`. Scoped broadly for any project shell script tests, not just CI plumbing.
- New `pnpm test:bats` script in root `package.json` for local test runs.
- Documentation: CHANGELOG not needed (internal CI infrastructure). README not needed.

## Key Discoveries

- `grep -oP` is not cross-platform — macOS grep lacks `-P`. Switched to `perl -nle` which works on both macOS and Linux.
- `|| echo "0"` fallback was broken — without pipefail, `tail -1` succeeds on empty input (exit 0) so `||` never fires. Replaced with `${VAR:-0}`.
- `Tests:.*\K\d+(?= total)` greedily consumed leading digits, matching only the trailing digit (1889 → 9). Fixed with `\s` boundary: `Tests:.*\s\K\d+(?= total)`.
- GitHub Actions composite actions resolve `$GITHUB_ACTION_PATH` to the action's directory at runtime, so scripts are referenced via relative path `$GITHUB_ACTION_PATH/../../scripts/ci/`.
- `resolve-qa-labels.js` already had a manual YAML parser and label filtering — extending it was simpler than maintaining a separate Python script with PyYAML.

## Test Plan

- [x] All existing Jest tests pass (`pnpm test`)
- [x] 12 bats tests pass (`bats tests/shell/`)
- [x] Formatting clean (`pnpm format:fix`)
- [ ] CI: `test-shell-scripts.yml` job runs successfully on PR

## Related

- Closes https://github.com/couimet/rangeLink/issues/572
- Follow-up issues: #574 (bats for test-release-run.sh), #575 (bats for validate-qa-coverage.sh)

## Release Notes

* **New Features**
  * Integration test failure counts now appear in PR CI summary comments
  * Failed test re-run commands are automatically extracted and displayed in PR comments for easier debugging

* **Chores**
  * Refactored CI infrastructure for improved test statistics extraction and reporting
  * Enhanced QA label resolution with additional filtering capabilities

[Review Change Stack](https://app.coderabbit.ai/change-stack/couimet/rangeLink/pull/576)

### CODE REVIEW

File: .github/workflows/test-shell-scripts.yml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+permissions:
+  contents: read
+
+jobs:
+  test:
+    runs-on: ubuntu-24.04
+    timeout-minutes: 5
+    steps:
+      - uses: actions/checkout@v4
+      - uses: bats-core/bats-action@4.0.0
```
[the remark this answers is not in the document]
Comment (couimet) [reply .github/workflows/test-shell-scripts.yml:17]: @coderabbitai I have https://github.com/couimet/rangeLink/issues/565 prepared to tackle this. I'll punt on it for now

[the remark this answers is not in the document]
Comment (couimet) [reply .github/workflows/test-shell-scripts.yml:17]: @coderabbitai Can you add a new comment on https://github.com/couimet/rangeLink/issues/565 with all the places we believe we should pin GitHub Actions to commit SHAs (not mutable tags).

Make sure to include the ones from this PR as well even if it's not yet on `main`
